src/HexagonOptimize.cpp

```diff
@@ -162,6 +162,8 @@ std::vector<Pattern> casts = {
     // Saturating narrowing casts
     { "halide.hexagon.satub.shr.vh.h", u8c(wild_i16x/wild_i16), Pattern::DeinterleaveOps | Pattern::ExactLog2Op1 },
     { "halide.hexagon.satuh.shr.vw.w", u16c(wild_i32x/wild_i32), Pattern::DeinterleaveOps | Pattern::ExactLog2Op1 },
+    { "halide.hexagon.sath.shr.vw.w", i16c(wild_i32x/wild_i32), Pattern::DeinterleaveOps | Pattern::ExactLog2Op1 },
+    { "halide.hexagon.satub.shr.vw.w", u8c(wild_i32x/wild_i32), Pattern::DeinterleaveOps | Pattern::ExactLog2Op1 },
```

Comment:
Looking at it again, I suspect the problem is with Pattern::DeinterleaveOps not being able to express the needed deinterleavings. You really need two deinterleavings, one at each half-narrowing.